File: Week 6/Sentimental-Credit/credit.py
def main():
    # Prompt user for a number
    number = input("Number: ")

    # Copies number into array
    arr = []
    chksum = []
    l = len(number)
    for i in range(l):
        digit = int(number[i])
        arr.append(digit)
        chksum.append(digit)

    # Calculates the values for the checksum
    x = l - 2
    while x >= 0:
        if chksum[x] > 4:
            prod = chksum[x] * 2
            chksum[x] = (prod % 10) + 1
        else:
            chksum[x] *= 2
        x -= 2

    # Adds up the checksum
    sum = 0
    for i in range(l):
        sum += chksum[i]

    # Checks if checksum is valid, prints invalid if otherwise
    if (sum % 10) == 0:
        # Prints name of card pattern number matches
        a = arr[0]
        b = arr[1]
        if amex(l, a, b) is True:
            print("AMEX")
        elif mastercard(l, a, b) is True:
            print("MASTERCARD")
        elif visa(l, a) is True:
            print("VISA")
        else:
            print("INVALID1")
    else:
        print("INVALID2")


# Checks if number matches Amex and returns true, false if otherwise
def  amex(len, x, y):
    if len == 15:
        if x == 3:
            if y == 4 or y == 7:
                return True
    return False


# Checks if number matches Mastercard and returns true, false if otherwise
def mastercard(len, x, y):
    if len == 16:
        if x == 5:
            if y >= 1 and y <= 5:
                return True
    return False


# Checks if number matches Visa and returns true, false if otherwise
def visa(len, x):
    if len == 13 or len == 16:
        if x == 4:
            return True
    return False


main()

File: Week 1/Credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

bool amex(int len, int a, int b);
bool mastercard(int len, int a, int b);
bool visa(int len, int x);

int main(void)
{
    // Declaring Variables.
    int len = 0;
    string num;

    // Prompts user for Card Number.
    num = get_string("Number: ");
    len = strlen(num);

    // Converts string to array of integers.
    for (int i = 0; i < len; i++)
    {
        num[i] -= 48;
    }
    // Creates An Array Of The Checksum Digits From Card Array.
    int chksum[len];
    for (int x = 0; x < len; x++)
    {
        chksum[x] = num[x];
    }
    int x = (len - 2);
    while (x >= 0)
    {
        if (chksum[x] > 4)
        {
            int prod = chksum[x] * 2;
            chksum[x] = (prod % 10) + 1;
        }
        else
        {
            chksum[x] = chksum[x] * 2;
        }
        x--;
        x--;
    }

    // Adds Up The Checksum Of The Digits.
    int total = 0;
    for (int y = 0; y < len; y++)
    {
        total = total + chksum[y];
    }

    // Prints Out The Name And Validity Of The Card.
    if ((total % 10) == 0)
    {
        int a = num[0];
        int b = num[1];
        if (amex(len, a, b) == true)
        {
            printf("AMEX\n");
        }
        else if (visa(len, a) == true)
        {
            printf("VISA\n");
        }
        else if (mastercard(len, a, b) == true)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

bool amex(int len, int a, int b)
{
    if (len == 15)
    {
        if (a == 3)
        {
            if ((b == 4) || (b == 7))
            {
                return true;
            }
        }
    }
    return false;
}

bool mastercard(int len, int a, int b)
{
    if (len == 16)
    {
        if (a == 5)
        {
            if ((b >= 1) && (b <= 5))
            {
                return true;
            }
        }
    }
    return false;
}

bool visa(int len, int x)
{
    if ((len == 13) || (len == 16))
    {
        if (x == 4)
        {
            return true;
        }
    }
    return false;
}
